x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/assignment/planning/LinearProgrammingPlanSolver.java

```diff
@@ -0,0 +1,341 @@
+
+        Tuple<Map<Tuple<Model, Node>, Double>, AssignmentPlan> weightsAndBinPackingPlan = calculateWeightsAndBinPackingPlan();
+
+        Map<Tuple<Model, Node>, Double> allocationValues = new HashMap<>();
+        Map<Tuple<Model, Node>, Double> assignmentValues = new HashMap<>();
+        if (solveLinearProgram(weightsAndBinPackingPlan.v1(), allocationValues, assignmentValues) == false) {
+            return weightsAndBinPackingPlan.v2();
+        }
+
+        RandomizedAssignmentRounding randomizedRounding = new RandomizedAssignmentRounding(
+            random,
+            RANDOMIZED_ROUNDING_ROUNDS,
+            nodes,
+            models
+        );
+        AssignmentPlan assignmentPlan = randomizedRounding.computePlan(allocationValues, assignmentValues);
+        AssignmentPlan binPackingPlan = weightsAndBinPackingPlan.v2();
+        if (binPackingPlan.compareTo(assignmentPlan) > 0) {
+            assignmentPlan = binPackingPlan;
+            logger.debug(() -> "Best plan is from bin packing");
+        } else {
+            logger.debug(() -> "Best plan is from LP solver");
+        }
+
+        return assignmentPlan;
+    }
+
+    private double weightForAllocationVar(Model m, Node n, Map<Tuple<Model, Node>, Double> weights) {
+        return (1 + weights.get(Tuple.tuple(m, n)) - (m.memoryBytes() > n.availableMemoryBytes() ? 10 : 0)) - L1 * normalizedMemoryPerModel
+            .get(m) / maxNodeCores;
+    }
+
+    private Tuple<Map<Tuple<Model, Node>, Double>, AssignmentPlan> calculateWeightsAndBinPackingPlan() {
+        logger.debug(() -> "Calculating weights and bin packing plan");
+
+        double w = INITIAL_W;
+        double dw = w / nodes.size() / models.size();
+
+        Map<Tuple<Model, Node>, Double> weights = new HashMap<>();
+        AssignmentPlan.Builder assignmentPlan = AssignmentPlan.builder(nodes, models);
+
+        for (Model m : models.stream().sorted(Comparator.comparingDouble(this::dsafModelOrder)).toList()) {
+            while (true) {
+                List<Node> orderedNodes = nodes.stream()
+                    .sorted(Comparator.comparingDouble(n -> dsafNodeOrder(n, m, assignmentPlan)))
+                    .toList();
+                double lastW = w;
+                for (Node n : orderedNodes) {
+                    int allocations = Math.min(
+                        assignmentPlan.getRemainingCores(n) / m.threadsPerAllocation(),
+                        assignmentPlan.getRemainingAllocations(m)
+                    );
+                    if (allocations > 0 && assignmentPlan.canAssign(m, n, allocations)) {
+                        assignmentPlan.assignModelToNode(m, n, allocations);
+                        weights.put(Tuple.tuple(m, n), w);
+                        w -= dw;
+                        break;
+                    }
+                }
+                if (lastW == w || assignmentPlan.getRemainingAllocations(m) == 0) {
+                    break;
+                }
+            }
```

Comment:
```suggestion
            double lastW = w;
            do {
                List<Node> orderedNodes = nodes.stream()
                    .sorted(Comparator.comparingDouble(n -> dsafNodeOrder(n, m, assignmentPlan)))
                    .toList();
                for (Node n : orderedNodes) {
                    int allocations = Math.min(
                        assignmentPlan.getRemainingCores(n) / m.threadsPerAllocation(),
                        assignmentPlan.getRemainingAllocations(m)
                    );
                    if (allocations > 0 && assignmentPlan.canAssign(m, n, allocations)) {
                        assignmentPlan.assignModelToNode(m, n, allocations);
                        weights.put(Tuple.tuple(m, n), w);
                        w -= dw;
                        break;
                    }
                }
            } while(lastW != w && assignmentPlan.getRemainingAllocations(m) != 0);
```

Seems safer than having a `while(true)` to me